File: src/p2xxx/p2419.py
from leetcode_prelude import *


# Problem 2419
class Solution:
    def longestSubarray(
        self,
        nums: List[int],
    ) -> int:
        maximum = 0
        running_length = 1
        max_length = 1

        prev = -1

        for idx, item in enumerate(nums):
            if item > maximum:
                # new maximum, reset the running length and the maximum length
                maximum = item
                running_length = 1
                max_length = 1
            elif item == maximum:
                # same maximum, bump running length if possible
                # update new the max length
                if item == prev:
                    running_length += 1
                else:
                    running_length = 1
                max_length = max(max_length, running_length)
            else:
                # otherwise, running_length is set to zero
                # but it is also handled in the equality case
                pass

            prev = item

        return max_length

File: src/p2xxx/p2419.rs
// Problem 2419
impl Solution {
    pub fn longest_subarray(nums: Vec<i32>) -> i32 {
        let mut longest = 1;

        let mut prev = 0;
        let mut max = 0;
        let mut len = 1;

        for item in nums.iter().copied() {
            match item.cmp(&max) {
                std::cmp::Ordering::Less => {}
                std::cmp::Ordering::Equal => {
                    if item == prev {
                        len += 1;
                        longest = longest.max(len);
                    } else {
                        len = 1;
                    }
                }
                std::cmp::Ordering::Greater => {
                    max = item;
                    len = 1;
                    longest = 1;
                }
            }

            prev = item;
        }

        longest
    }
}
